packages/entity/src/utils/EntityPrivacyUtils.ts

```diff
@@ -0,0 +1,311 @@
+ * is insufficient for the task at hand. When dealing with purely a sequence of mutations it is easy
+ * to roll back all mutations given a single authorization error by wrapping them in a single transaction.
+ * When certain portions of a mutation cannot be rolled back transactionally (third pary calls,
+ * legacy code, etc), using this method can help decide whether the sequence of mutations will fail before
+ * attempting them. Note that if any privacy policy rules use a piece of data being updated in the mutations
+ * the result of this method and the update mutation itself may differ.
+ *
+ * @param entityClass - class of entity
+ * @param sourceEntity - entity loaded by viewer
+ * @param queryContext - query context in which to perform the check
+ */
+export async function canViewerUpdateAsync<
+  TMFields extends object,
+  TMID extends NonNullable<TMFields[TMSelectedFields]>,
+  TMViewerContext extends ViewerContext,
+  TMEntity extends Entity<TMFields, TMID, TMViewerContext, TMSelectedFields>,
+  TMPrivacyPolicy extends EntityPrivacyPolicy<
+    TMFields,
+    TMID,
+    TMViewerContext,
+    TMEntity,
+    TMSelectedFields
+  >,
+  TMSelectedFields extends keyof TMFields = keyof TMFields
+>(
+  entityClass: IEntityClass<
+    TMFields,
+    TMID,
+    TMViewerContext,
+    TMEntity,
+    TMPrivacyPolicy,
+    TMSelectedFields
+  >,
+  sourceEntity: TMEntity,
+  queryContext: EntityQueryContext = sourceEntity
+    .getViewerContext()
+    .getViewerScopedEntityCompanionForClass(entityClass)
+    .getQueryContextProvider()
+    .getQueryContext()
+): Promise<boolean> {
+  return await canViewerUpdateInternalAsync(
+    entityClass,
+    sourceEntity,
+    /* cascadingDeleteCause */ null,
+    queryContext
+  );
+}
+
+async function canViewerUpdateInternalAsync<
+  TMFields extends object,
+  TMID extends NonNullable<TMFields[TMSelectedFields]>,
+  TMViewerContext extends ViewerContext,
+  TMEntity extends Entity<TMFields, TMID, TMViewerContext, TMSelectedFields>,
+  TMPrivacyPolicy extends EntityPrivacyPolicy<
+    TMFields,
+    TMID,
+    TMViewerContext,
+    TMEntity,
+    TMSelectedFields
+  >,
+  TMSelectedFields extends keyof TMFields = keyof TMFields
+>(
+  entityClass: IEntityClass<
+    TMFields,
+    TMID,
+    TMViewerContext,
+    TMEntity,
+    TMPrivacyPolicy,
+    TMSelectedFields
+  >,
+  sourceEntity: TMEntity,
+  cascadingDeleteCause: EntityCascadingDeletionInfo | null,
+  queryContext: EntityQueryContext
+): Promise<boolean> {
+  const companion = sourceEntity
+    .getViewerContext()
+    .getViewerScopedEntityCompanionForClass(entityClass);
+  const privacyPolicy = companion.entityCompanion.privacyPolicy;
+  const evaluationResult = await asyncResult(
+    privacyPolicy.authorizeUpdateAsync(
+      sourceEntity.getViewerContext(),
+      queryContext,
+      { cascadingDeleteCause },
+      sourceEntity,
+      companion.getMetricsAdapter()
+    )
+  );
+  return evaluationResult.ok;
```

Comment:
Shouldn't we check the error type like we check it for deletes? Specifically, if the database is down or the code is buggy ("undefined is not an object") we should rethrow instead of returning false.